lib/queryTicket.js

```diff
@@ -1,61 +1,43 @@
-import fetch from 'node-fetch'
-import { parseString } from 'xml2js'
 import { parseLegs, price } from './parse.js'
-
-const endpoint = 'https://fahrkarten.bahn.de/mobile/dbc/xs.go?'
+import requestBackend from './requestBackend.js';
 
 /**
  * queryTicket function
  *
  * @param   string | DB ticket number (Auftragsnummer)
  * @param   lastName | last name of purchaser 
+ * @param   kwid | kwid of the ticket (only needed for tickets from the new system)
  * @return  Promise | fulfills with an object that contains the ticket data
  */
-const queryTicket = async (ticketNumber, lastName) => {
-
-  const reqBody = '<?xml version="1.0"?><rqorderdetails version="1.0">'
-    + '<rqheader tnr="61743782011" ts="2022-07-15T22:29:00" l="de" v="22060000" d="iPhone13,1" os="iOS_15.5" app="NAVIGATOR"/>'
-    + '<rqorder on="' + ticketNumber + '"/><authname tln="' + lastName + '"/></rqorderdetails>'
+const queryTicket = async (ticketNumber, lastName,kwid=null) => {
 
-  return fetch(endpoint, {
-    method: 'POST',
-    headers: {
-      'User-Agent': 'github.com/envake/db-tickets',
-      'Accept-Language': 'de-DE,de;q=0.9'
-    },
-    body: reqBody
-  })
-  .then(response => {
-    if (!response.ok) {
-      throw new HTTPResponseError(response)
-    }
-    return response.text()
-  })
-  .then(body => {
-    return new Promise((resolve, reject) => parseString(body, (err, result) => {
-      if (err) {
-        reject(err)
-      }
-      else {
-        resolve(result)
-      }
-    }))
-  })
-  .then(json => {
-    if (json['rperror']) {
-      throw new XMLRPCError(json['rperror']['error'])
+  let reqBody = '<?xml version="1.0"?><rqorderdetails version="2.0">'
+    + '<rqheader l="de" v="23080000" d="iPhone16.4.1" os="iOS_15.7.5" app="NAVIGATOR"/>';
+    if(kwid){
+      reqBody += '<rqorder on="' + ticketNumber + '" kwid="' + kwid + '"/>';
+    }else{
+      reqBody += '<rqorder on="' + ticketNumber + '"/>';
     }
+     reqBody +='<authname tln="' + lastName + '"/></rqorderdetails>';
 
+  return requestBackend(reqBody)
+  .then(json => {
+  
     const data = json.rporderdetails
-    console.log(JSON.stringify(data))
+    
   
     // selectors
     const _order = data['order'][0]['$']
     const _tck = data['order'][0]['tcklist']?.[0]['tck'][0]
     const _ticket = _tck?.['mtk'][0]
+    let _encoded_price = _tck['htdata'][0]['ht'][0]['_']
+    //if _tck['htdata'][0]['ht'][0]['_'] starts with data:image/png;base64, then it is a old ticket, we need to check for that
```

Comment:
I think this is a very intricate detail. Shall we add a test case for it, to make sure parsing both ticket variants won't fail in the future?

Reply:
Good idea, feel free to do so, I'm actually not a JS Dev so I'm not feeling completely comfortable to bring a unknown testing framework to the library.

Reply:
Because apparently I'm not allowed to push into this PR's source branch, I have pushed my changes to [`pr-5` within this repo](https://github.com/public-transport/db-tickets/tree/pr-5).

I have added a basic test case with a modified (so that it doesn't contain my personal data, including the base64-encoded barcode & "Sichtmerkmal"!) ticket.
If you have an old-style ticket (or API response for it, rather) at hand, please check it in and add a second test case.

So far, I haven't used any fancy test framework, as we can always introduce more abstractions & tooling later. `test/index.js` merely uses [Node.js built-in `assert` module](https://nodejs.org/docs/latest-v16.x/api/assert.html). From Node v18 on, it also includes [a basic test runner](https://nodejs.org/docs/latest-v18.x/api/test.html), but I don't think that we need it.

Reply:
Implemented an old ticket :)